Update testing.md (#1800)
zio/zio: docs/usecases/testing.md

The library also includes built-in mock versions of all the standard ZIO environmental effects (`Clock`, `Console`, `System`, and `Random`), so when we tested our program above the `helloWorld` method didn't actually print a String to the console but instead wrote the String to a buffer that could access for testing purposes. If you ever do need to access the "live" environment just use the `live` method in the `mock` package or specify the live environment in your type signature like `Live[Console]`.

_Mocking Random_

> When working with randomness it is easier to controll what the radnom producer supplies. This can be controlled by a number of methods that allow for filling up buffers with values.

Reviewer: ```suggestion
When working with randomness it is easier to control what the random producer supplies. This can be controlled by a number of methods that allow for filling up buffers with values.
```

```scala
import zio.test.mock.MockRandom
testM("One can provide its own list of ints") {
  for {
    _  <- MockRandom.feedInts(1, 9, 2, 8, 3, 7, 4, 6, 5)
    r1 <- random.nextInt
    r2 <- random.nextInt
    r3 <- random.nextInt
    r4 <- random.nextInt
    r5 <- random.nextInt
    r6 <- random.nextInt
    r7 <- random.nextInt
    r8 <- random.nextInt
    r9 <- random.nextInt
  } yield assert(
    List(1, 9, 2, 8, 3, 7, 4, 6, 5),
    equalTo(List(r1, r2, r3, r4, r5, r6, r7, r8, r9))
  )
}
```

_Mocking Clock_

> Sometimes one need to be able to control the flow of time. In most cases you want your unit test to be as fast as possible. Waiting for readl time to pass by is a real killer for this. ZIO exposes a MockClock in testing Environment that can controll time.

Reviewer: ```suggestion
Sometimes one need to be able to control the flow of time. In most cases you want your unit test to be as fast as possible. Waiting for real time to pass by is a real killer for this. ZIO exposes a MockClock in testing Environment that can control time.
```

Example 1

> Below test will pass. If we woule remove the line `MockClock.adjust(Duration(10, TimeUnit.SECONDS))` it would fail. If we wouldn't have access to `MockClock` to make it pass we would need to put manual `ZIO.sleep` call and that test would take 10 seconds every time when run.

Reviewer: ```suggestion
Below test will pass. If we would remove the line `MockClock.adjust(Duration(10, TimeUnit.SECONDS))` it would fail. If we wouldn't have access to `MockClock` to make it pass we would need to put manual `ZIO.sleep` call and that test would take 10 seconds every time when run.
```

```scala
import zio.test.mock.MockClock
testM("One can move time very fast") {  
  for {
    startTime <- currentTime(TimeUnit.SECONDS)
    _         <- MockClock.adjust(Duration(10, TimeUnit.SECONDS))
    endTime   <- currentTime(TimeUnit.SECONDS)
  } yield assert(endTime - startTime, isGreaterThanEqualTo(10L))
}
```

Example 2

`MockClock` affects also all code running asynchronously that is scheduled to run after certain time but with caveats to how runtime works.

```scala
testM("One can controll time as he see fit") {
  for {
    ref     <- Ref.make(0)
    _       <- (ZIO.sleep(Duration(10, TimeUnit.SECONDS)) *> ref.update(_+1)).fork
    _       <- MockClock.adjust(Duration(11, TimeUnit.SECONDS)) 
    readRef <- ref.get
  } yield assert(1, equalTo(readRef))
}
```

> The above code cretes a shared mutable cell that will be incremented by 1 after 10 seconds. With call to `MockClock.adjust(Duration(11, TimeUnit.SECONDS))` we simulate passing of 11 seconds of time. Unfortunately *there is no gurantee* that the ref is already increamented. `ZIO.sleep` only guarantees that at least 10 seconds will pass. There is no upper bound. This depends on when the runtime decides to run fiber that control process of updating this cell.

Reviewer: ```suggestion
The above code creates a shared mutable cell that will be incremented by 1 after 10 seconds. With call to `MockClock.adjust(Duration(11, TimeUnit.SECONDS))` we simulate passing of 11 seconds of time. Unfortunately *there is no guarantee* that the ref is already incremented. `ZIO.sleep` only guarantees that at least 10 seconds will pass. There is no upper bound. This depends on when the runtime decides to run fiber that control process of updating this cell.
```

> The above code cretes a shared mutable cell that will be incremented by 1 after 10 seconds. With call to `MockClock.adjust(Duration(11, TimeUnit.SECONDS))` we simulate passing of 11 seconds of time. Unfortunately *there is no gurantee* that the ref is already increamented. `ZIO.sleep` only guarantees that at least 10 seconds will pass. There is no upper bound. This depends on when the runtime decides to run fiber that control process of updating this cell.

Author: Are you sure about `will "wake up" in order of earliest time first`? E.g. you would put two fibers to sleep and there would be third that was evaluated by the runtime longer than both sleep took. Isn't the runtime free to choose which fiber to pickup first for further evaluation?

> The above code cretes a shared mutable cell that will be incremented by 1 after 10 seconds. With call to `MockClock.adjust(Duration(11, TimeUnit.SECONDS))` we simulate passing of 11 seconds of time. Unfortunately *there is no gurantee* that the ref is already increamented. `ZIO.sleep` only guarantees that at least 10 seconds will pass. There is no upper bound. This depends on when the runtime decides to run fiber that control process of updating this cell.

Reviewer 2: Good point. The guarantee we can provide is that we will always submit the tasks to the runtime in order of earliest time first (see [here](https://github.com/zio/zio/blob/9d5601de8adf1471ddfd49a0c364175b70d0b285/test/shared/src/main/scala/zio/test/mock/MockClock.scala#L263). So if you have three tasks that are supposed to wake in 9 seconds, 10 seconds, and 11 seconds and you advance the clock by 12 seconds, we will send the first one to the runtime, then the second, and then the third. But you're right that we are sending them to the runtime in that order but almost simultaneously and it is up to the runtime at that point what to do with them.

At the end we wait on the promise until it is set. With call to `MockClock.adjust(10.seconds)` we simulate passing of 10 seconds of time.
Because of it we don't need to wait for the real 10 seconds to pass and thus our unit test can run faster. This is a pattern that will very often be used when `sleep` and `MockClock` are being used for testing of effects that are based on time. The fiber that needs to sleep will be forked and `MockClock` will used to adjust the time so that all expected effects are run in the forked fiber.

*WARNING* 
Notice that if we don't call `adjust` at all we'll get stuck. `MockClock` doesn't make any progress on its own.

Also it is worth mentioning that adjusting the time on `MockClock` doesn't make us immune to the timing overheads introduced by the runtime and races this introduces. Effects are guaranteed to be waken up not earlier than the argument passed to `sleep` but from there the order of execution since scheduled on different threads is indeterministic and its up to the user code to use tools like `Promise` to guarantee proper sequencing. 

> Below code will fail because there is non-zero overhead when switching fibers thus reading of the value happens before it is set and its change is 

Reviewer 2: You may want to say this test will be flaky versus it will fail. Since there is a race condition between getting and setting the `Ref` the test may pass or fail each time it is run but it will not reliably pass which is what we want for our tests.

Author: Right 👌

Because of it we don't need to wait for the real 10 seconds to pass and thus our unit test can run faster. This is a pattern that will very often be used when `sleep` and `MockClock` are being used for testing of effects that are based on time. The fiber that needs to sleep will be forked and `MockClock` will used to adjust the time so that all expected effects are run in the forked fiber.

*WARNING* 
Notice that if we don't call `adjust` at all we'll get stuck. `MockClock` doesn't make any progress on its own.

Also it is worth mentioning that adjusting the time on `MockClock` doesn't make us immune to the timing overheads introduced by the runtime and races this introduces. Effects are guaranteed to be waken up not earlier than the argument passed to `sleep` but from there the order of execution since scheduled on different threads is indeterministic and its up to the user code to use tools like `Promise` to guarantee proper sequencing. 
Below code will be flaky because there is non-zero overhead when switching fibers thus reading of the value might happen before it is set and its change is 
propagated.

```scala

> testM("THIS TEST WILL FAIL - Sleep and adjust can introduce races") {

Reviewer 3: Would be nice to show how to fix this by using a `Promise` that is set _after_ the `sleep` inside the forked fiber; then awaiting on the promise right before `MockClock.adjust`.

Author: Isn't it exactly this example?
https://github.com/zio/zio/pull/1800/files#diff-22f248e1a7bb07e86d37efa910d22132R190

more or less the same amount of time. Biggest source of complexity during testing comes from external services which we cannot control like external
payment APIs, object storages, http APIs ect. It is normal to hide these kind of services behind an interface and provide mock implementation to regain
control and determinism. However there is another source of complexity that comes from the local infrastructure that is also hard to control without building prior abstractions. Things like stdin/stdout, clocks, random generators, schedulers can make writing tests hard or even impossible. Fortunately ZIO abstracted most of it in its runtime under `Environment` type. Thanks to this design `zio-test` could easily provide its own implementation named `MockEnvironment` which gives you mocks of mentioned infrastructure. In most of the cases when you'll be using `ZIO`s `testM` mocks are already created and should be controlled by exposed functions on companion object. If for some reason you would like to provide custom environment or are using other testing framework but still want to use mocks there are `make` functions on companion objects of mocks where you can construct your own. 

> A mistake that is easy to do is to create new instance of mock but use the one provided by the `MockEnvironment`. 

Reviewer 2: Maybe say accidentally use different mock instances at the same time?

Author: 🆗

**Zero Dependencies**

As a library with zero third party dependencies, this project is available on the JVM, ScalaJS, and will be available on Dotty and Scala Native in the near future. So you can write your tests once and make sure that your code works correctly across all platforms that you support.

**How zio-test was designed**

> `zio-test` is designed around the idea of making test first-class objects. What it means is that tests 

Reviewer 2: I think this is definitely right. I might say that not only are tests values, they are also effects. This has a couple of advantages. First, we solve the asynchrony problem where in other frameworks you have to "run" your effects to test them, but then you can't block, especially on Javascript, so then you basically need to rewrap them in some new effect or quasi-effect type like `Future`, creating a lot of indirection. Second, making tests effects gives us access to all the tools of `ZIO` which lets use solve lots of problems in very simple and straightforward ways. Want a timeout on your tests? You don't need special logic. `ZIO` already has a `timeout` method`. Retry a test a bunch of times to make sure it isn't flaky? `repeat`? Use some type of fixture? `Managed`.

Author: 👍

(and other accompanying concepts like assertions) become ordinary values that can be passed around, transformed and composed together. This approach allows for greater flexibility comparing to some other testing frameworks where tests and additional logic around tests had to be put into callbacks so that framework could make use of them. This approach also fits better with other `ZIO` concepts like `ZManaged` which can only be used within a scoped block of code. This also created a mismatch between `BeforeAll`, `AfterAll` callback-like methods when there were resources that should be opened and closed during test suite execution.

_Constructing tests_

All below code assumes that you have imported `zio.test._`

The backbone of `zio-test` is the `Spec[L, T]` class. Every spec is labeled with `L` and can be a suite which contains other specs or a test of type `T`.

> The most common and easy way to create suits and tests is to use the `suite` and `testM` functions. 

Reviewer 2: You might want to mention `test` for testing pure functions and `testM` for testing effectual functions.

Author: 👍

```scala
import zio.test._
import zio.clock.nanoTime
import Assertion.isGreaterThan

val clockSuite = suite("clock") {
  testM("time is non-zero") {
    assertM(nanoTime, isGreaterThan(0))
  }
}
```

> As you can see the whole suit was assigned to `clockSuite` val. As it was said suits can contain other suits so we can aggregate them as much as needed. Example, we can have multiple suits that test external HTTP apis and one big suit that will aggregate them all.

Reviewer 2: "Suites". I think I saw that a couple of other places so would check for spelling throughout.

```scala
val paymentProviderABCSuite  = suit("ABC payment provider tests") {/* tests */}
val paymentProviderXYZSuite  = suit("XYZ payment provider tests") {/* tests */}
val allPaymentProvidersTests = suit("All payment providers tests")(paymentProviderABCSuite, paymentProviderXYZSuite)
```

Real tests that run some logic and return testing result are created mostly with `testM` function. It expects two arguments, first one that will be the label of test
which will be used for visual reporting back to the user and an assertion of type
`ZIO[R, E, TestResult]`. This means writing test in `zio-test` mostly gets down to creating a `ZIO` object that 
will produce `TestResult`. There is another variant of function for creating test that are pure called simply `test`.
It expects a thunk of code that will just return a `TestResult` without packing it into `ZIO`.

_Assertions - creating TestResults_

> As it was already mentioned tests should return `TestResult`. The most common way to produce a `TestResult`

Reviewer 2: I think this is really good. The other thing I might add here is how a lot of assertions compose by taking other assertions as parameters. For example you could write `isRight(isSome(equalTo(2)))`. The idea with this and using methods like `&&` and `||` is that a lot of examples for testing frameworks have these really simply tests where you are comparing two integers or something, so if the test fails it is obvious what went wrong. But in real life maybe you are comparing part of two large case classes, or a bunch of different statements about some test. So it isn't that helpful to just say this test failed. We want to really zero in for you on what part of your test failed so it is easy for you to diagnose the problem and fix it.

Author: 👍

**Zero Dependencies**

As a library with zero third party dependencies, this project is available on the JVM, ScalaJS, and will be available on Dotty and Scala Native in the near future. So you can write your tests once and make sure that your code works correctly across all platforms that you support.

**How zio-test was designed**

> `zio-test` is designed around the idea of making test first-class objects. What it means is that tests (and other accompanying concepts like assertions) become ordinary values that can be passed around, transformed and composed together. This approach allows for greater flexibility comparing to some other testing frameworks where tests and additional logic around tests had to be put into callbacks so that framework could make use of them. This approach also fits better with other `ZIO` concepts like `ZManaged` which can only be used within a scoped block of code. This also created a mismatch between `BeforeAll`, `AfterAll` callback-like methods when there were resources that should be opened and closed during test suite execution.

Reviewer 2: "tests" instead of "test" in the first sentence.

and at best wrap them in `scala.util.Future` because blocking would eliminate running on ScalaJS, `zio-test` expects you to create `ZIO` objects. There is no need for indirect transformations from one wrapping object to another. Second, because our tests are ordinary `ZIO` values we don't need to turn to testing framework for things like retries, timeouts and resource management. We can solve all those problems with full richness of functions that `ZIO` exposes.

_Constructing tests_

All below code assumes that you have imported `zio.test._`

The backbone of `zio-test` is the `Spec[L, T]` class. Every spec is labeled with `L` and can be a suite which contains other specs or a test of type `T`.

The most common and easy way to create suites is to use `suite` function. For testing of pure functions there is `test` function and for effectful testing there is `testM`

```scala
import zio.test._
import zio.clock.nanoTime
import Assertion.isGreaterThan

val clockSuite = suite("clock") {
  testM("time is non-zero") {
    assertM(nanoTime, isGreaterThan(0))
  }
}
```

> As you can see the whole suit was assigned to `clockSuite` val. As it was said suites can contain other suites so we can aggregate them as much as needed. Example, we can have multiple suites that test external HTTP apis and one big suit that will aggregate them all.

Reviewer 4: ```suggestion
As you can see the whole suit was assigned to `clockSuite` val. As it was said suites can contain other suites so we can aggregate them as much as needed. Example, we can have multiple suites that test external HTTP apis and one big suite that will aggregate them all.
```

```scala

> val paymentProviderABCSuite  = suit("ABC payment provider tests") {/* tests */}

Reviewer 4: ```suggestion
val paymentProviderABCSuite  = suite("ABC payment provider tests") {/* tests */}
```

> val paymentProviderXYZSuite  = suit("XYZ payment provider tests") {/* tests */}

Reviewer 4: ```suggestion
val paymentProviderXYZSuite  = suite("XYZ payment provider tests") {/* tests */}
```

> val allPaymentProvidersTests = suit("All payment providers tests")(paymentProviderABCSuite, paymentProviderXYZSuite)

Reviewer 4: ```suggestion
val allPaymentProvidersTests = suite("All payment providers tests")(paymentProviderABCSuite, paymentProviderXYZSuite)
```

which will be used for visual reporting back to the user and an assertion of type
`ZIO[R, E, TestResult]`. This means writing test in `zio-test` mostly gets down to creating a `ZIO` object that 
will produce `TestResult`. There is another variant of function for creating test that are pure called simply `test`.
It expects a thunk of code that will just return a `TestResult` without packing it into `ZIO`.

_Assertions - creating TestResults_
As it was already mentioned tests should return `TestResult`. The most common way to produce a `TestResult`
is to resort to `assert` or its effectful counterpart `assertM`. Both of them accept a value of type `A` (effectful version wrapped in a `ZIO`) and an `Assertion[A]`.
To create `Assertion[A]` object one can use functions defined under `zio.test.Assertion`. There are already a number
of useful assertions predefined like `equalTo`, `isFalse`, `isTrue`, `contains`, `throws` and more.
What is really useful in assertions is that they behave like boolean values and can be composed with operators

> known from operating on boolean values like and (`&&`), or (`||`), negation (`negate`)

Reviewer 4: .
```suggestion
known from operating on boolean values like and (`&&`), or (`||`), negation (`negate`).
```

Here the expression `Right(Some(2))` is of type `Either[Any, Option[Int]]`and our assertion `isRight(isSome(equalTo(2)))`
is of type `Assertion[Either[Any, Option[Int]]]`

```scala
import zio.test.Assertion.{isRight, isSome,equalTo, isGreaterThanEqualTo, not, hasField}

final case class Address(country:String, city:String)
final case class User(name:String, age:Int, address: Address)

test("Rich checking") {
  assert(
    User("Jonny", 26, Address("Denmark", "Copenhagen")),
    hasField("age", (u:User) => u.age, isGreaterThanEqualTo(18)) &&
    hasField("country", (u:User) => u.address.country, not(equalTo("USA")))
  )
}
```
There is also an easy way to test object's data for certain assertions with `hasField` which accepts besides a name, a mapping function from object to its tested property and `Assertion` object which will validate this property. Here our test checks if a person has at least 18 years and is not from USA. What is nice about those tests is that, test reporters will tell you exactly which assertion was broken. Let's say we would change `isGreaterThanEqualTo(18)` to `isGreaterThanEqualTo(40)` which will fail. Printout

> on console will be a nice detailed text what exactly went wrong:

Reviewer 4: ```suggestion
on console will be a nice detailed text explaining what exactly went wrong:
```

```bash
[info]       User(Jonny,26,Address(Denmark,Copenhagen)) did not satisfy (hasField("age", _.age, isGreaterThanEqualTo(45)) && hasField("country", _.country, not(equalTo(USA))))
[info]       26 did not satisfy isGreaterThanEqualTo(45)
```

Having this all in mind probably the most common and also most readable way of structuring tests is to pass
a for-comprehension to `testM` function and yield a call to `assert` function.

```scala
testM("Semaphore should expose available number of permits") {
  for {
    s         <- Semaphore.make(1L)
    permits   <- s.available
  } yield assert(permits, equalTo(1L))
}
```

_Running tests_

> When all of our test are constructed we need to have a way to actually execute them. Your first stop is the `zio.test.DefaultRunnableSpec` which accepts a single suite that will be executed. A single suite might seem to be limiting but as it was already said suit can hold any number of other suites. You may structure your tests like this:

Reviewer 4: ```suggestion
When all of our tests are constructed, we need to have a way to actually execute them. Your first stop is the `zio.test.DefaultRunnableSpec` which accepts a single suite that will be executed. A single suite might seem to be limiting but as it was already said suites can hold any number of other suites. You may structure your tests like this:
```

```scala
val suite1 = suite("suite1") {
  testM("s1.t1") {...},
  testM("s1.t2") {...}
}
val suite2 = suite("suite2") {
  testM("s2.t1") {...},
  testM("s2.t2") {...},
  testM("s2.t3") {...}
}
val suite3 = suite("suite3") {
  testM("s3.t1") {...}
}

object AllSuites extends DefaultRunnableSpec(suite("All tests")(suite1, suite2, suite3))
```

`DefaultRunnableSpec` is very similar in its logic of operations to `zio.App`. Instead of providing one `ZIO` application

> at the end of the world we provide a suite that can be a tree of other suites and tests. Another resemblance is that `DefaultRunnableSpec` provides an Environment. Here it is an instance of `MockEnvironment` which helps us with controling our systems infrastructure. More info on using mocks can be found in sections bellow. 

Reviewer 2: "below" instead of "bellow".

Just like with `zio.App` where at the very end an instance of `ZIO[R,E,A]` is expected where `R` can be at maximum of type `Environment` in `DefaultRunnableSpec` `R` cannot be more than `MockEnvironment`. So just like in normal application if our
`R` is composed of some other modules we need to provide them first before test can be executed. How can we provide our dependencies?

> Here again the design of `zio-test` shines. Since our tests are ordinary values we can jest transform them with call to `mapTest`.

Reviewer 4: ```suggestion
Here again the design of `zio-test` shines. Since our tests are ordinary values we can just transform them with a call to `mapTest`.
```

> It accepts a lambda of type `ZIO[R with MockSystem, TestFailure[Throwable], TestSuccess[Unit] ] => T1`. Wihtout getting into too much details about types we can see that our lambda argument is a test instance (`ZIO`) that expects an environment of type `R with MockSystem`. This is no different from normal usage of ZIO in `zio.App`. We can use the same `provide`, `provideSome` methods to provide modules which `DefaultRunnableSpec` cannot provide itselt as those are users modules.

Reviewer 4: ```suggestion
It accepts a lambda of type `ZIO[R with MockSystem, TestFailure[Throwable], TestSuccess[Unit] ] => T1`. Without getting into too much details about types we can see that our lambda argument is a test instance (`ZIO`) that expects an environment of type `R with MockSystem`. This is no different from normal usage of ZIO in `zio.App`. We can use the same `provide`, `provideSome` methods to provide modules which `DefaultRunnableSpec` cannot provide itself as those are users modules.
```

> 

Reviewer 2: Maybe talk a little more about how you can either run specs extending `DefaultRunnableSpec` with the main method using `test:run`` or with the SBT test runner that will automatically detect all your tests and run them?

_Using Mocks in tests_

What we expect from tests (at least those that we consider unit tests) is to be stable i.e. consecutive runs should yield the same results and take
more or less the same amount of time. Biggest source of complexity during testing comes from external services which we cannot control like external

> payment APIs, object storages, http APIs ect. It is normal to hide these kind of services behind an interface and provide mock implementation to regain

Reviewer 4: ```suggestion
payment APIs, object storages, http APIs etc. It is normal to hide these kind of services behind an interface and provide mock implementation to regain
```

        permits   <- s.available
      } yield assert(permits, equalTo(2L))
    }
```

Above code doesn't work. We created a new `MockClock` instance and are correctly adjusting its time. What might be surprising is that call to `timeout` will
use the `MockClock` provided by the `MockEnvironment` not our `mockClock` instance. It easy to know why when you look at the signature of `timeout`:

```scala
final def timeout(d: Duration): ZIO[R with Clock, E, Option[A]] 
```

The returned type is `ZIO[R with Clock, E, Option[A]]` where our environment is "some R plus a Clock".
Before running this `Clock` has to be provided and the framework provides the Clock from the `MockEnvironment` not our instance variable as it is not aware that we created it.

If you need to provide real implementations instead of the mocks to some part of your tests there is a `live` method which will transform your `ZIO[R, E, A]` to `ZIO[Live[R], E, A]`. Going from `R` to `Live[R]` instructs the framework that we really want to be provided with live implementations and not mocks.

If you have a non-trivial flow of data from multiple streams that can produce at different intervals and would like to test

> snapshots of data in particular point in time a nice pattern is to join these streams with `zipWithLatest` and 

Reviewer 2: I would generalize this a little bit. `zipWithLatest` is the specific method we are testing here but it is a more general pattern. If we have a single value we're waiting for after a certain amount of time we can have the effect complete a promise, advance the clock, and wait on the promise. But if we want to advance time multiple times and wait on multiple results, using a queue gives us the same ability to "wait" for the next result without having to create multiple promises.

Author: OK, will talk about it and I think I will move this example to MockClock section. It makes more sense there.
